File: Tp map/GMapActivity/app/src/main/java/ma/ensa/gmapactivity/MainActivity.java
package ma.ensa.gmapactivity;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.location.LocationProvider;
import android.os.Bundle;
import android.provider.Settings;
import android.telephony.TelephonyManager;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

public class MainActivity extends AppCompatActivity {
    private double latitude;
    private double longitude;
    private double altitude;
    private float accuracy;
    RequestQueue requestQueue;
    Button mapButton;
    String insertUrl = "http://192.168.102.74/geometrique/createPosition.php";
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });
        mapButton = findViewById(R.id.button);
        requestQueue = Volley.newRequestQueue(getApplicationContext());
        mapButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(MainActivity.this, MapsActivity.class);
                startActivity(intent);
            }
        });
        LocationManager locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);

        if (ActivityCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                && ActivityCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 10000, 150, new LocationListener() {
            @Override
            public void onStatusChanged(String provider, int status, Bundle extras) {
                String newStatus = "";
                switch (status) {
                    case LocationProvider.OUT_OF_SERVICE:
                        newStatus = "OUT_OF_SERVICE";
                        break;
                    case LocationProvider.TEMPORARILY_UNAVAILABLE:
                        newStatus = "TEMPORARILY_UNAVAILABLE";
                        break;
                    case LocationProvider.AVAILABLE:
                        newStatus = "AVAILABLE";
                        break;
                }
                String msg = String.format(getResources().getString(R.string.provider_new_status), provider, newStatus);
                Toast.makeText(getApplicationContext(), msg, Toast.LENGTH_SHORT).show();
            }
            @Override
            public void onProviderEnabled(String provider) {
                String msg = String.format(getResources().getString(R.string.provider_enabled), provider);
                Toast.makeText(getApplicationContext(), msg, Toast.LENGTH_SHORT).show();
            }
            @Override
            public void onProviderDisabled(String provider) {
                String msg = String.format(getResources().getString(R.string.provider_disabled), provider);
                Toast.makeText(getApplicationContext(), msg, Toast.LENGTH_SHORT).show();
            }
            @Override
            public void onLocationChanged(Location location) {
                latitude = location.getLatitude();
                longitude = location.getLongitude();
                altitude = location.getAltitude();
                accuracy = location.getAccuracy();
                Log.d("Location Update", "Latitude: " + latitude + ", Longitude: " + longitude);
                addPosition(latitude, longitude);
                Toast.makeText(getApplicationContext(), "Location Updated: Lat: " + latitude + ", Lon: " + longitude, Toast.LENGTH_LONG).show();
            }
            void addPosition(final double lat, final double lon) {
                StringRequest request = new StringRequest(Request.Method.POST, insertUrl,
                        new Response.Listener<String>() {
                            @Override
                            public void onResponse(String response) {
                                // Log response from server
                                Log.d("Server Response", "Response from server: " + response);
                                Toast.makeText(getApplicationContext(), "Server response: " + response, Toast.LENGTH_SHORT).show();
                            }
                        }, new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        Log.e("Server Error", "Error sending data: " + error.getMessage());
                        Toast.makeText(getApplicationContext(), "Error sending data: " + error.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                }) {
                    @Override
                    protected Map<String, String> getParams() {
                        HashMap<String, String> params = new HashMap<>();
                        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
                        params.put("latitude", String.valueOf(lat));
                        params.put("longitude", String.valueOf(lon));
                        params.put("date", sdf.format(new Date()));
                        String androidId = Settings.Secure.getString(getContentResolver(), Settings.Secure.ANDROID_ID);
                        params.put("imei", androidId);
                        Log.d("Sending Params", "Params: " + params.toString());
                        return params;
                    }
                };
                requestQueue.add(request);
            }











            private void testDatabaseInsert() {
                StringRequest request = new StringRequest(Request.Method.POST, insertUrl,
                        new Response.Listener<String>() {
                            @Override
                            public void onResponse(String response) {
                                Log.d("Test Insert", "Response from server: " + response);
                                Toast.makeText(getApplicationContext(), "Test insert successful: " + response, Toast.LENGTH_SHORT).show();
                            }
                        }, new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        Log.e("Test Insert Error", "Error: " + error.toString());
                        Toast.makeText(getApplicationContext(), "Test insert failed: " + error.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                }) {
                    @Override
                    protected Map<String, String> getParams() throws AuthFailureError {
                        HashMap<String, String> params = new HashMap<>();
                        params.put("latitude", "30.12345"); // Random test latitude
                        params.put("longitude", "-97.12345"); // Random test longitude
                        params.put("date", "2024-10-25 10:00:00"); // Random date
                        params.put("imei", "testIMEI123456"); // Random IMEI for testing
                        return params;
                    }
                };

                requestQueue.add(request);
            }
        });
    }
}

File: permission/app/src/main/java/ma/ensa/project/MainActivity.kt
package ma.ensa.project

import android.Manifest
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.graphics.Insets
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

class MainActivity() : AppCompatActivity() {

    private val PERMISSION_RECORD_AUDIO = Manifest.permission.RECORD_AUDIO
    private val PERMISSION_ACCESS_FINE_LOCATION = Manifest.permission.ACCESS_FINE_LOCATION
    private val PERMISSION_CAMERA = Manifest.permission.CAMERA

    private val PERMISSION_REQ_CODE_AUDIO = 100
    private val PERMISSION_REQ_CODE_LOCATION = 101
    private val PERMISSION_REQ_CODE_CAMERA = 102



    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById<View>(R.id.main)
        ) { v: View, insets: WindowInsetsCompat ->
            val systemBars: Insets = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        findViewById<View>(R.id.buttonaudio).setOnClickListener {
            requestRuntimePermission(PERMISSION_RECORD_AUDIO, PERMISSION_REQ_CODE_AUDIO, "audio")
        }

        findViewById<View>(R.id.buttonlocation).setOnClickListener {
            requestRuntimePermission(PERMISSION_ACCESS_FINE_LOCATION, PERMISSION_REQ_CODE_LOCATION, "location"
            )
        }

        findViewById<View>(R.id.buttoncamera).setOnClickListener {
            requestRuntimePermission(PERMISSION_CAMERA, PERMISSION_REQ_CODE_CAMERA, "camera"
            )
        }

        updatePermissionStatus()
    }

    private fun updatePermissionStatus() {
        // Check if permissions are granted and update TextView for each one
        val audioResult = findViewById<TextView>(R.id.audioresult)
        val locationResult = findViewById<TextView>(R.id.Locationresult)
        val cameraResult = findViewById<TextView>(R.id.cameraresult)

        if (checkSelfPermission(PERMISSION_RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED) {
            audioResult.text = "Permission to Audio is ON"
        } else {
            audioResult.text = "Permission to Audio is OFF"
        }

        if (checkSelfPermission(PERMISSION_ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            locationResult.text = "Permission to Location is ON"
        } else {
            locationResult.text = "Permission to Location is OFF"
        }

        if (checkSelfPermission(PERMISSION_CAMERA) == PackageManager.PERMISSION_GRANTED) {
            cameraResult.text = "Permission to Camera is ON"
        } else {
            cameraResult.text = "Permission to Camera is OFF"
        }
    }

    private fun requestRuntimePermission(
        permission: String,
        requestCode: Int,
        featureName: String
    ) {
        if (checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED) {
            Toast.makeText(this, "$featureName permission is granted", Toast.LENGTH_LONG).show()
        } else if (ActivityCompat.shouldShowRequestPermissionRationale(this, permission)) {
            val builder = AlertDialog.Builder(this)
            builder.setMessage("This app requires $featureName permission for this feature")
                .setTitle("Permission Request")
                .setCancelable(false)
                .setPositiveButton("OK") { dialog: DialogInterface, which: Int ->
                    ActivityCompat.requestPermissions(
                        this@MainActivity,
                        arrayOf(permission),
                        requestCode
                    )
                    dialog.dismiss()
                }
                .setNegativeButton("Cancel") { dialog: DialogInterface, which: Int -> dialog.dismiss() }
            builder.show()
        } else {
            ActivityCompat.requestPermissions(this, arrayOf(permission), requestCode)
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            when (requestCode) {
                PERMISSION_REQ_CODE_AUDIO -> Toast.makeText(
                    this, "Audio Permission Granted", Toast.LENGTH_SHORT
                ).show()

                PERMISSION_REQ_CODE_LOCATION -> Toast.makeText(this, "Location Permission Granted", Toast.LENGTH_SHORT
                ).show()

                PERMISSION_REQ_CODE_CAMERA -> Toast.makeText(this, "Camera Permission Granted", Toast.LENGTH_SHORT
                ).show()
            }
        } else {
            val featureName = if (requestCode == PERMISSION_REQ_CODE_AUDIO) "Audio"
            else if (requestCode == PERMISSION_REQ_CODE_LOCATION) "Location"
            else "Camera"
            showPermissionDeniedDialog(featureName)
        }
        updatePermissionStatus()
    }

    private fun showPermissionDeniedDialog(featureName: String) {
        val builder = AlertDialog.Builder(this)
        builder.setMessage(
            "This feature is unavailable because " + featureName + " permission has been denied. " +
                    "Please allow this permission from settings."
        )
            .setTitle("Permission Required")
            .setCancelable(false)
            .setPositiveButton("Settings") { dialog: DialogInterface, which: Int ->
                val intent: Intent =
                    Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                val uri: Uri = Uri.fromParts("package", getPackageName(), null)
                intent.setData(uri)
                startActivity(intent)
                dialog.dismiss()
            }
            .setNegativeButton("Cancel"
            ) { dialog: DialogInterface, which: Int -> dialog.dismiss() }
        builder.show()
    }
}